### PR TITLE
Tobin descriptor unique binding error enum

### PR DESCRIPTION
**Pull request by tobine**
Updates to CreateDescriptorSetLayout:

1. Split into Pre/Post sections to perform validation prior to calling down chain
2. Add unique error enum
3. Add validation flag
4. Add test for duplicate binding error

### CODE REVIEW

File: layers/descriptor_sets.cpp [hunk 1]
```diff
@@ -24,18 +24,13 @@
...
     : layout_(layout), binding_count_(p_create_info->bindingCount), descriptor_count_(0), dynamic_descriptor_count_(0) {
     uint32_t global_index = 0;
     for (uint32_t i = 0; i < binding_count_; ++i) {
         descriptor_count_ += p_create_info->pBindings[i].descriptorCount;
-        if (!binding_to_index_map_.emplace(p_create_info->pBindings[i].binding, i).second) {
-            log_msg(report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_DESCRIPTOR_SET_LAYOUT_EXT,
-                    reinterpret_cast<uint64_t &>(layout_), __LINE__, VALIDATION_ERROR_02345, "DS",
-                    "duplicated binding number in VkDescriptorSetLayoutBinding. %s", validation_error_map[VALIDATION_ERROR_02345]);
-        }
+        binding_to_index_map_.emplace(p_create_info->pBindings[i].binding, i);
```
Comment (chrisforbes) [inline layers/descriptor_sets.cpp]: This can probably just use `operator[]`. The only reason it used `emplace` was to get feedback on the collision.

File: layers/descriptor_sets.cpp [hunk 2]
```diff
@@ -53,6 +48,23 @@ cvdescriptorset::DescriptorSetLayout::DescriptorSetLayout(debug_report_data *rep
...
     }
 }
+
+// Validate descriptor set layout create info
+bool cvdescriptorset::DescriptorSetLayout::ValidateCreateInfo(debug_report_data *report_data,
+                                                              const VkDescriptorSetLayoutCreateInfo *create_info) {
+    bool skip = false;
+    std::unordered_set<uint32_t> bindings;
+    for (uint32_t i = 0; i < create_info->bindingCount; ++i) {
+        if (bindings.count(create_info->pBindings[i].binding)) {
```
Comment (chrisforbes) [inline layers/descriptor_sets.cpp]: I /would/ use `insert`'s return value here

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -6715,17 +6715,32 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateSampler(VkDevice device, const VkSamplerCre
...
 
+static bool PreCallValidateCreateDescriptorSetLayout(layer_data *dev_data, const VkDescriptorSetLayoutCreateInfo *create_info) {
+    if (dev_data->instance_data->disabled.create_descriptor_set_layout)
+        return false;
+    return cvdescriptorset::DescriptorSetLayout::ValidateCreateInfo(dev_data->report_data, create_info);
+}
+
+static void PostCallRecordCreateDescriptorSetLayout(layer_data *dev_data, const VkDescriptorSetLayoutCreateInfo *create_info,
+                                                    VkDescriptorSetLayout set_layout) {
+    // TODOSC : Capture layout bindings set
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: This comment is junk.

File: layers/core_validation.cpp [hunk 4]
```diff
@@ -6715,17 +6715,32 @@ VKAPI_ATTR VkResult VKAPI_CALL CreateSampler(VkDevice device, const VkSamplerCre
...
+    dev_data->descriptorSetLayoutMap[set_layout] = new cvdescriptorset::DescriptorSetLayout(create_info, set_layout);
```
Comment (chrisforbes) [inline layers/core_validation.cpp]: Still leaking these-- a problem for a follow-up PR maybe. Value type should be unique_ptr

**Comment by tobine** [comment]
Incorporated feedback and pushed
